chore: docs: Add more details for debugging datapath packet loss. (#12681)
magma/magma: docs/readmes/howtos/troubleshooting/datapath_connectivity.md

5. From here onwards you are going to debug OVS datapath, so you need to select
   a UE and identify which traffic direction is broken. You can do so by
   - Generating uplink traffic in UE

>    - Capture packets on gtpu_sys_2152 device: `tcpdump -eni gtpu_sys_2152 host $UE_IP`. If you do not see any packet, it means the packets are not reaching GTP tunnel, check S1 connectivity.

Reviewer: - As it belongs to the sentence starting with `You can do so by ...` it should be `capturing` imho.
- `it means that` would sound more like a sentence to me.
- The last comma should probably be a full stop followed by a new sentence [as short sentences are prefered](https://github.com/magma/magma/wiki/Contributing-Documentation#styling-and-writing-conventions).

>    - Capturing packets on gtpu_sys_2152 device: `tcpdump -eni gtpu_sys_2152 host $UE_IP`. If you do not see any packet, it means that packets are not reaching GTP tunnel. Check S1 connectivity to debug further.

Reviewer 2: ```suggestion
   - Capturing packets on gtpu_sys_2152 device: `tcpdump -eni gtpu_sys_2152 host $UE_IP`. If you do not see any packet, it means that packets are not reaching the GTP tunnel. Check S1 connectivity to debug further.
```

>    - *NATed datapath*: Capture packets on gtp_br0 `tcpdump -eni gtp_br0 host $UE_IP`. If you dont see any packets debug using `dp_probe_cli.py` utility. This utility would show which OVS table is dropping the packet.

Reviewer: Maybe "If you do not see any packets, debug using the `dp_probe_cli.py` utility."?

>    - *NATed datapath*: Capture packets on gtp_br0 `tcpdump -eni gtp_br0 host $UE_IP`. If you dont see any packets, try debugging with `dp_probe_cli.py` utility. This utility would show which OVS table is dropping the packet.

Reviewer 2: ```suggestion
   - *NATed datapath*: Capture packets on gtp_br0 `tcpdump -eni gtp_br0 host $UE_IP`. If you don't see any packets, try debugging with the `dp_probe_cli.py` utility. This utility would show which OVS table is dropping the packet.
```

>    - *NATed datapath*: you also need to check if packet are egressing on

Reviewer: `NATed datapath` is duplicated. Is this intentional? Might this be combined into one item?

Author: these are two separate steps, so I would like to keep it separate.

>    - *NATed datapath*: you also need to check if packet are egressing on

Reviewer: For consistency with the above, `You` should probably be capitalized.

Reviewer: As this line is duplicated, this holds also for the second occurrence 3 lines above. :wink:

>    - *NATed datapath*: you also need to check if packet are egressing on

Reviewer 2: ```suggestion
   - *NATed datapath*: you also need to check if the packet is egressing on
```

Reviewer 2: or "the packets are"

     the SGi port. You can do so by running tcpdump on SGi port

>      `tcpdump -eni $SGi_dev dst $SERVER_IP`. In case the packet is missing on SGi port, you have issue with the routing. Check routing table and iptables rules on the AGW.

Reviewer: - `an issue`
- `Check the routing table`

>      `tcpdump -eni $SGi_dev dst $SERVER_IP`. In case the packet is missing on SGi port, you have an issue with the routing. Check the routing table and iptables rules on the AGW.

Reviewer 2: ```suggestion
     `tcpdump -eni $SGi_dev dst $SERVER_IP`. In case the packet is missing on the SGi port, you have an issue with the routing. Check the routing table and iptables rules on the AGW.
```

>    - *Non-NAT datapath*: You also need to check if packet are egressing on

Reviewer 2: ```suggestion
   - *Non-NAT datapath*: You also need to check if the packet is egressing on
```

Reviewer 2: Same as above with "the packets are"

>      the SGi port. You can do so by running tcpdump on SGi port

Reviewer 2: ```suggestion
     the SGi port. You can do so by running tcpdump on the SGi port
```

>      `tcpdump -eni $SGi_dev dst $SERVER_IP`. If you dont see any packets debug using `dp_probe_cli.py` utility. This utility would show which OVS table is dropping the packet.

Reviewer: Duplication to the `NATed datapath` case. There should be a clearer distinction, what is applicable for both cases or only one.

Author: thats what the prefix (*Non-NAT datapath*, "*NATed datapath*") trying to do.

Reviewer: Let me maybe be more explicit here. The text is:

>    - *NATed datapath*: you also need to check if packet are egressing on
>      the SGi port. You can do so by running tcpdump on SGi port
>      `tcpdump -eni $SGi_dev dst $SERVER_IP`. In case the packet is missing on SGi port, you have an issue with the routing. Check the routing table and iptables rules on the AGW.
>    - *Non-NAT datapath*: You also need to check if packet are egressing on
>      the SGi port. You can do so by running tcpdump on SGi port
>      `tcpdump -eni $SGi_dev dst $SERVER_IP`. If you dont see any packets, try debugging with `dp_probe_cli.py` utility. This utility would show which OVS table is dropping the packet.

More than half of it is duplicated (3 lines):
```markdown
You also need to check if packet are egressing on
the SGi port. You can do so by running tcpdump on SGi port
`tcpdump -eni $SGi_dev dst $SERVER_IP`. 
```
Imho it would be better to avoid the confusing duplication of those parts and be clearer and more structured here. E.g. one could do:
```markdown
- You also need to check if packet are egressing on the SGi port. You can do so by running tcpdump
on the SGi port `tcpdump -eni $SGi_dev dst $SERVER_IP`.
  - *NATed datapath*: In case ...
  - *Non-NAT datapath*: If you do not ...
```

Author: Users are either going to debug NAT or Non-NAT datapath, so duplicate is not issue here.
The suggested structure looks more confusing to user who is looking to debug either NAT ot Non_NAT datapath.

Reviewer: In that case, that users look for either NAT or non_NAT debugging, we should than probably group by that aspect to make this easier for the user.

Author: That is worth doing it, but it would be separate task. I have created doc for this: https://github.com/magma/magma/issues/12719

>      `tcpdump -eni $SGi_dev dst $SERVER_IP`. If you dont see any packets, try debugging with `dp_probe_cli.py` utility. This utility would show which OVS table is dropping the packet.

Reviewer 2: ```suggestion
     `tcpdump -eni $SGi_dev dst $SERVER_IP`. If you don't see any packets, try debugging with the `dp_probe_cli.py` utility. This utility would show which OVS table is dropping the packet.
```
